src/Common/src/Interop/Linux/System.Net.Security.Native/Interop.Ntlm.cs

```diff
@@ -0,0 +1,124 @@
+
+        [DllImport(Interop.Libraries.SecurityNative, EntryPoint="NetSecurity_HeimNtlmDecodeType2")]
+        internal static extern int HeimNtlmDecodeType2(byte[] data, int offset, int count, out SafeNtlmType2Handle type2Handle);
+
+        [DllImport(Interop.Libraries.SecurityNative, EntryPoint="NetSecurity_HeimNtlmFreeType2")]
+        internal static extern int HeimNtlmFreeType2(IntPtr type2Handle);
+
+        [DllImport(Interop.Libraries.SecurityNative, EntryPoint="NetSecurity_HeimNtlmNtKey", CharSet = CharSet.Ansi)]
+        internal static extern int HeimNtlmNtKey(string password, out SafeNtlmBufferHandle key, out int length);
+
+        [DllImport(Interop.Libraries.SecurityNative, EntryPoint="NetSecurity_HeimNtlmCalculateResponse", CharSet = CharSet.Ansi)]
+        internal static extern int HeimNtlmCalculateResponse(
+            bool isLM,
+            SafeNtlmBufferHandle key,
+            SafeNtlmType2Handle type2Handle,
+            string username,
+            string target,
+            byte[] baseSessionKey,
+            int baseSessionKeyLen,
+            out SafeNtlmBufferHandle answer,
+            out int ansLength);
+
+        [DllImport(Interop.Libraries.SecurityNative, EntryPoint="NetSecurity_CreateType3Message", CharSet = CharSet.Ansi)]
+        internal static extern int CreateType3Message(
+            SafeNtlmBufferHandle key,
+            SafeNtlmType2Handle type2Handle,
+            string username,
+            string domain,
+            uint flags,
+            SafeNtlmBufferHandle lmResponse,
+            SafeNtlmBufferHandle ntlmResponse,
+            byte [] baseSessionKey,
+            int baseSessionKeyLen,
+            out SafeNtlmBufferHandle sessionKey,
+            out int sessionKeyLen,
+            out SafeNtlmBufferHandle data,
+            out int dataLen
+            );
+
+        internal partial class NtlmFlags
+        {
+            internal const uint NTLMSSP_NEGOTIATE_UNICODE = 0x1;
+            internal const uint NTLMSSP_REQUEST_TARGET = 0x4;
+            internal const uint NTLMSSP_NEGOTIATE_SIGN = 0x10;
+            internal const uint NTLMSSP_NEGOTIATE_SEAL = 0x20;
+            internal const uint NTLMSSP_NEGOTIATE_NTLM = 0x200;
+            internal const uint NTLMSSP_NEGOTIATE_ALWAYS_SIGN = 0x8000;
+            internal const uint NTLMSSP_NEGOTIATE_EXTENDED_SESSIONSECURITY = 0x80000;
+            internal const uint NTLMSSP_NEGOTIATE_128 = 0x20000000;
+            internal const uint NTLMSSP_NEGOTIATE_KEY_EXCH = 0x40000000;
+        }
+
+        internal static byte[] EVPDigest(byte[] key, byte[] input, int inputlen, out uint outputlen)
+        {
+            //reference doc: http://msdn.microsoft.com/en-us/library/cc236700.aspx
+            byte[] output = new byte[Interop.Crypto.EVP_MAX_MD_SIZE];
+            outputlen = 0;
+            using (SafeEvpMdCtxHandle ctx = Interop.Crypto.EvpMdCtxCreate(Interop.Crypto.EvpMd5()))
+            unsafe
```

Comment:
This looks like it's just an incremental hash of key and input... why not just use the IncrementalHash class?